addons/binding/org.openhab.binding.nikohomecontrol/src/main/java/org/openhab/binding/nikohomecontrol/internal/protocol/NikoHomeControlCommunication.java

```diff
@@ -0,0 +1,783 @@
+            this.value2 = value2;
+            this.value3 = value3;
+        }
+
+        public void setStartValue(Integer startValue) {
+            this.startValue = startValue;
+        }
+
+        public void setEndValue(Integer endValue) {
+            this.endValue = endValue;
+        }
+
+    }
+
+    /**
+     * Class {@link NHCCmdAck} used as output from gson for cmd or event feedback from Niko Home Control where the data
+     * part is enclosed by [] and contains a list of json strings.
+     * <p>
+     * Example: <code>{"cmd":"listactions","data":[{"id":1,"name":"Garage","type":1,"location":1,"value1":0},
+     * {"id":25,"name":"Frontdoor","type":2,"location":2,"value1":0}]}</code>
+     *
+     * @author Mark Herwege
+     */
+    private static class NHCCmdAck {
+        private String cmd;
+        private String event;
+        private ArrayList<HashMap<String, String>> data = new ArrayList<>();
+    }
+
+    /**
+     * Class {@link NHCSimpleCmdAck} used as output from gson for cmd or event feedback from Niko Home Control where the
+     * data part is a simple json string.
+     * <p>
+     * Example: <code>{"cmd":"executeactions", "data":{"error":0}}</code>
+     *
+     * @author Mark Herwege
+     */
+    private static class NHCSimpleCmdAck {
+        private String cmd;
+        @SuppressWarnings("unused")
+        private String event;
+        private HashMap<String, String> data = new HashMap<>();
+    }
+
+    /**
+     * Constructor for Niko Home Control communication object, manages communication with
+     * Niko Home Control IP-interface.
+     *
+     * @param addr Can be null or omitted, will attempt to discover IP address.
+     * @param port
+     */
+    public NikoHomeControlCommunication(InetAddress addr, int port) {
+
+        startCommunication(addr, port);
+    }
+
+    public NikoHomeControlCommunication(int port) {
+
+        startCommunication(null, port);
+    }
+
+    /**
+     * Get Niko Home Control IP-interface IP address.
+     * <p>
+     * The method sends a UDP packet with content 0x44 to the local network on port 10000.
+     * The Niko Home Control IP-interface responds to this UDP packet.
+     * The IP-address from the Niko Home Control IP-interface is then extracted from the response packet.
+     *
+     * @param IP address used to get local broadcast address, will be broadcast in local subnet if null
+     * @return IP address
+     * @throws IOException
+     */
+    private InetAddress nikoHomeControlFindAddr(InetAddress broadcastaddr) throws IOException {
+
+        byte[] discoverbuffer = { (byte) 0x44 };
+        int broadcastport = 10000;
+
+        if (broadcastaddr == null) {
+            broadcastaddr = InetAddress.getLocalHost();
+        }
+        byte[] ipaddr = broadcastaddr.getAddress();
+        ipaddr[3] = (byte) 0xff;
+        broadcastaddr = InetAddress.getByAddress(ipaddr);
+
+        DatagramSocket datagramSocket = null;
+        try {
```

Comment:
You can use the try with resources on the `DatagramSocket` then you can get rid of the finally:
https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html